Sprint 12

desmoo0 opened the pull request.

desmoo0 committed:
Update README.md

desmoo0 committed:
Add Friendship status

desmoo0 committed:
Fix FriendStorage

desmoo0 committed:
Fix FriendStorage #2

desmoo0 committed:
Fix FriendStorage #3

desmoo0 committed:
Fix tests

desmoo0 committed:
Fix tests #2

desmoo0 committed:
Fix tests #3

desmoo0 committed:
Bugfixes

desmoo0 committed:
Bugfixes #2

desmoo0 committed:
Bugfixes #3

desmoo0 committed:
Bugfixes #4

desmoo0 committed:
Bugfixes #5

desmoo0 committed:
Bugfixes #6

desmoo0 committed:
Sprint 12

desmoo0 committed:
Update 2

desmoo0 committed:
Delete db directory

desmoo0 committed:
Fix checkstyle

desmoo0 committed:
Fix tests checkstyle

desmoo0 committed:
Fix checkstyle #2

desmoo0 committed:
Fix checkstyle #3

desmoo0 committed:
Fix checkstyle #4

desmoo0 committed:
Delete src/main/java/ru/yandex/practicum/filmorate/controller/DictController.java

desmoo0 committed:
Fix tests #4

desmoo0 committed:
Fix tests #5

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/film/Film.java:
```diff
@@ -0,0 +1,56 @@
+        if (rating == null) {
+            this.mpa = null;
+        } else {
+            Mpa m = new Mpa(rating.getId());
```
Давай избавимся от всех однобуквенных названий переменных и сокращений, они делают код менее читаемым

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/film/Genre.java:
```diff
@@ -0,0 +1,41 @@
+        this.name = name;
+    }
+
+    public static final Genre COMEDY       = new Genre(1, "Комедия");
```
В рамках тз нет определенного списка жанров. Пользователь должен мочь добавлять любое значение

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/film/MpaRating.java:
```diff
@@ -0,0 +1,37 @@
+
+@Getter
+public enum MpaRating {
+    G(1, "G"),
```
В соответствии с тз нет определенного списка рейтингов. Пользователь должен мочь добавить любое значение

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,34 +1,56 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
 public class Film {
```
Этот объект дублирует объект из папки model/fim. Нужно оставит только один. Удали также дублирующие объекты Genre, Mpa, User

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,70 +1,97 @@
-        Film film = findById(filmId);
-        userService.findById(userId);
-        film.getLikes().add(userId);
+    public FilmService(@Qualifier("filmDbStorage") FilmStorage films,
```
Нет смысла добавлять к каждому бину `@Qualifier`, спринг может подставить корректное значение и без него, если у этого бина всего одна реализация. Давай удалим лишние `@Qualifier` здесь и в сервисах ниже.

desmoo0 replied:
<img width="1431" height="252" alt="image" src="https://github.com/user-attachments/assets/fd687413-3a7f-4b59-a778-2fcf5406eba9" />
Я оставлю только на filmDbStorage @Qualifier

desmoo0 replied:
и userDbStorage

afomkina replied:
Нужно удалить все InMemory реализации, так как они нам больше не понадобятся

desmoo0 replied:
Готово!

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/FriendStorage.java:
```diff
@@ -0,0 +1,18 @@
+import ru.yandex.practicum.filmorate.model.friend.Friendship;
+import java.util.List;
+
+public interface FriendStorage {
```
Нигде не используется, нужно использовать

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFriendStorage.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.*;
+
+@Component
+public class InMemoryFriendStorage implements FriendStorage {
```
Все inmemory реализации нужно удалить. Они больше не будут использоваться в коде. В коммерчесткой разработке не принято хранить неиспользуемый код, так как хранится вся история изменений в гите

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,131 @@
+    public List<User> getCommonFriends(long userId, long otherId) {
+        final String sql = "SELECT u.* FROM USERS u WHERE u.ID IN (" +
+                "  SELECT FRIEND_ID FROM FRIENDS WHERE USER_ID = ? " +
+                "  INTERSECT " +
```
Вложенные подзапросы и INTERSECT утяжеляют план и усложняют индексацию. Нужно переписать запрос используя джоины.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,131 @@
+    public List<User> findAll() {
+        final String sql = "SELECT ID, EMAIL, LOGIN, NAME, BIRTHDAY FROM USERS ORDER BY ID";
+        return jdbc.query(sql, (rs, rn) -> {
+            User u = new User();
```
Дублируется RowMapper<User> в нескольких методах. Вынеси его в приватное поле RowMapper<User> USER_MAPPER = (rs, rn) -> { ... } и переиспользуй. Это уменьшит риски расхождений и упростит поддержку. Поправь во всех репозиториях.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+        try {
+            jdbc.update("INSERT INTO LIKES (FILM_ID, USER_ID) VALUES (?, ?)", filmId, userId);
+        } catch (DuplicateKeyException ignore) {
+            // Игнорируем попытку повторно поставить лайк, потом можно будет реализовать убирание лайка
```
Лучше не оставлять пустой catch, можно тут логировать ошибки, это поможет с дебагом, если что-то пойдет не так

desmoo0 committed:
Delete src directory

desmoo0 committed:
Update fix

desmoo0 committed:
Bugfixes

desmoo0 committed:
Fix checkstyle #5

desmoo0 committed:
Fix checkstyle #6

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -17,38 +20,50 @@ public class ErrorHandler {
-        );
-        log.warn("Ошибка валидации: {}", errors);
-        return new ValidationErrorResponse("Ошибка валидации", errors);
+        ex.getBindingResult().getFieldErrors().forEach(err -> errors.put(err.getField(), err.getDefaultMessage()));
```
Из за множественных сокращений и однобуквенных переменных код сложно читать. Нудно придерживаться code-style при написании кода. В данной работе я отмечу все места, где названия переменных не соответсвуют code-style. Нужно будет их все переменовать.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -17,38 +20,50 @@ public class ErrorHandler {
 
     @ResponseStatus(HttpStatus.NOT_FOUND)
     @ExceptionHandler(NoSuchElementException.class)
     public ErrorResponse handleNotFound(NoSuchElementException ex) {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -17,38 +20,50 @@ public class ErrorHandler {
 
     @ResponseStatus(HttpStatus.BAD_REQUEST)
     @ExceptionHandler(ExistingUserException.class)
     public ErrorResponse handleUserExists(ExistingUserException ex) {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -17,38 +20,50 @@ public class ErrorHandler {
 
     @ResponseStatus(HttpStatus.BAD_REQUEST)
     @ExceptionHandler(ExistingMovieException.class)
     public ErrorResponse handleFilmExists(ExistingMovieException ex) {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -17,38 +20,50 @@ public class ErrorHandler {
-    public ErrorResponse handleAny(Exception ex) {
-        log.error("Необработанная ошибка: {}", ex.getMessage(), ex);
-        return new ErrorResponse("Внутренняя ошибка сервера: " + ex.getMessage());
+    public ErrorResponse handleAnyException(Exception ex) {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -17,38 +20,50 @@ public class ErrorHandler {
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler(NotFoundException.class)
+    public ErrorResponse handleNotFound(NotFoundException ex) {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -17,38 +20,50 @@ public class ErrorHandler {
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(SQLException.class)
+    public ErrorResponse handleDbError(SQLException ex) {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,70 +1,97 @@
+                .orElseThrow(() -> new NotFoundException("MPA " + film.getMpa().getId())));
+
+        Set<Genre> resolvedGenres = film.getGenres().stream()
+                .map(g -> {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+
+        jdbc.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(sql, new String[]{"ID"});
```
Нужно переименовать переменную

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+        final String sql = "SELECT F.ID, F.NAME, F.DESCRIPTION, F.RELEASE_DATE, F.DURATION, F.MPA_ID, M.NAME as MPA_NAME " +
+                "FROM FILM AS F JOIN MPA AS M ON F.MPA_ID = M.ID ORDER BY F.ID";
+        List<Film> films = jdbc.query(sql, this::mapRowToFilm);
+        Map<Long, Film> filmsById = films.stream().collect(Collectors.toMap(Film::getId, f -> f));
```
- Нужно переименовать параметр f
- Стримы принято бить по строкам для лучшей читаемости

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+                "ORDER BY COUNT(DISTINCT l.USER_ID) DESC, f.ID ASC " +
+                "LIMIT ?";
+        List<Film> films = jdbc.query(sql, this::mapRowToFilm, count);
+        Map<Long, Film> filmsById = films.stream().collect(Collectors.toMap(Film::getId, f -> f));
```
- Нужно переименовать параметр f
- Стримы принято бить по строкам для лучшей читаемости

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+        return films;
+    }
+
+    private Film mapRowToFilm(ResultSet rs, int rowNum) throws SQLException {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+        }
+
+        List<Object[]> batch = genres.stream()
+                .map(g -> new Object[]{filmId, g.getId()})
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+        try {
+            jdbc.batchUpdate("INSERT INTO FILM_GENRE (FILM_ID, GENRE_ID) VALUES (?, ?)", batch);
+            log.info("Successfully synced {} genres for film {}", genres.size(), filmId);
+        } catch (Exception e) {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+
+        log.debug("Loading genres for {} films", filmsById.size());
+
+        jdbc.query(sql, rs -> {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/MpaDbStorage.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public Optional<Mpa> findById(int id) {
+        final String sql = "SELECT ID, NAME FROM MPA WHERE ID=?";
+        List<Mpa> list = jdbc.query(sql, (rs, rn) ->
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/MpaDbStorage.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public List<Mpa> findAll() {
+        final String sql = "SELECT ID, NAME FROM MPA ORDER BY ID";
+        return jdbc.query(sql, (rs, rn) -> new Mpa(rs.getInt("ID"), rs.getString("NAME")));
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+
+    private final JdbcTemplate jdbc;
+    private final RowMapper<User> userMapper = (rs, rn) -> {
+        User u = new User();
```
Нужно переименовать переменную

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+public class UserDbStorage implements UserStorage {
+
+    private final JdbcTemplate jdbc;
+    private final RowMapper<User> userMapper = (rs, rn) -> {
```
Нужно переименовать параметры

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    public boolean containsKey(Long id) {
+        Integer cnt = jdbc.queryForObject("SELECT COUNT(1) FROM USERS WHERE ID=?", Integer.class, id);
```
Нужно переименовать переменную

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+    public Optional<User> findById(Long id) {
+        final String sql = "SELECT ID, EMAIL, LOGIN, NAME, BIRTHDAY FROM USERS WHERE ID=?";
+        List<User> list = jdbc.query(sql, (rs, rn) -> {
+            User u = new User();
```
Здесь нужно использовать rowMapper, который определен на 24 строке

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+    @Override
+    public Optional<User> findById(Long id) {
+        final String sql = "SELECT ID, EMAIL, LOGIN, NAME, BIRTHDAY FROM USERS WHERE ID=?";
+        List<User> list = jdbc.query(sql, (rs, rn) -> {
```
Нужно переименовать параметры

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+    @Override
+    public User create(User user) {
+        final String sql = "INSERT INTO USERS (EMAIL, LOGIN, NAME, BIRTHDAY) VALUES (?, ?, ?, ?)";
+        KeyHolder kh = new GeneratedKeyHolder();
```
Нужно переименовать переменную

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+    public User create(User user) {
+        final String sql = "INSERT INTO USERS (EMAIL, LOGIN, NAME, BIRTHDAY) VALUES (?, ?, ?, ?)";
+        KeyHolder kh = new GeneratedKeyHolder();
+        jdbc.update(con -> {
```
Нужно переименовать параметр

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+        final String sql = "INSERT INTO USERS (EMAIL, LOGIN, NAME, BIRTHDAY) VALUES (?, ?, ?, ?)";
+        KeyHolder kh = new GeneratedKeyHolder();
+        jdbc.update(con -> {
+            PreparedStatement ps = con.prepareStatement(sql, new String[]{"ID"});
```
Нужно переименовать переменную

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+        final String sql = "SELECT u.* FROM FRIENDS f " +
+                "JOIN USERS u ON u.ID = f.FRIEND_ID " +
+                "WHERE f.USER_ID = ? ORDER BY u.ID";
+        return jdbc.query(sql, (rs, rowNum) -> new User(
```
Здесь нужно использовать rowMapper, который опеределен на 24 строке

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+                " JOIN FRIENDS f1 ON u.ID = f1.FRIEND_ID AND f1.USER_ID = ? " +
+                " JOIN FRIENDS f2 ON u.ID = f2.FRIEND_ID AND f2.USER_ID = ? " +
+                " ORDER BY u.ID; ";
+        return jdbc.query(sql, (rs, rowNum) -> new User(
```
Нужно переиспользовать rowMapper

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorageIT.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Test
+    void create_find_update_and_containsKey() {
+        Film f = new Film();
```
Нужно переименовать переменную

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorageIT.java:
```diff
@@ -0,0 +1,111 @@
+        f.setReleaseDate(LocalDate.of(2001, 1, 1));
+        f.setDuration(100);
+        f.setMpa(new Mpa(1, null));
+        Set<Genre> g = new LinkedHashSet<>();
```
Нужно переименовать переменную

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorageIT.java:
```diff
@@ -0,0 +1,111 @@
+    @Test
+    void findAll_and_popular() {
+        // фильм 1
+        Film a = new Film();
```
Нужно переименовать переменную

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorageIT.java:
```diff
@@ -0,0 +1,111 @@
+        a = films.create(a);
+
+        // фильм 2
+        Film b = new Film();
```
Нужно переименовать переменную

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/GenreDbStorageIT.java:
```diff
@@ -0,0 +1,61 @@
+    JdbcTemplate jdbcTemplate;
+
+    @Autowired
+    GenreDbStorage genreDbStorage;
```
У полей класса не хватает модификаторов доступа

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/GenreDbStorageIT.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Test
+    void findById_missing_returnsEmpty() {
+        Optional<ru.yandex.practicum.filmorate.model.Genre> g = genreDbStorage.findById(999);
```
Нужно переименовать переменную
- `ru.yandex.practicum.filmorate.model.` нужно вынести в импорты

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/GenreDbStorageIT.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Test
+    void findById_existing_returnsGenre() {
+        Optional<ru.yandex.practicum.filmorate.model.Genre> g = genreDbStorage.findById(3);
```
Нужно переименовать переменную
- `ru.yandex.practicum.filmorate.model.` нужно вынести в импорты

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/GenreDbStorageIT.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Test
+    void findAll_returnsAllGenres_sortedByIdAsc() {
+        List<ru.yandex.practicum.filmorate.model.Genre> genres = genreDbStorage.findAll();
```
- `ru.yandex.practicum.filmorate.model.` нужно вынести в импорты

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/db/MpaDbStorageIT.java:
```diff
@@ -0,0 +1,40 @@
+        List<Mpa> all = storage.findAll();
+        assertThat(all).isNotEmpty();
+
+        Mpa m = storage.findById(1).orElseThrow();
```
Нужно переименовать переменную

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/UserDbStorageIT.java:
```diff
@@ -0,0 +1,65 @@
+
+    @BeforeEach
+    void setUp() {
+        // ничего не делаем
```
Нужно удалить `@BeforeEach`, если в нем ничего не происходит

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/UserDbStorageIT.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Test
+    void create_and_findById_and_containsKey() {
+        User u = new User(0L, "user@mail.ru", "login", "Имя", LocalDate.of(1990, 1, 1));
```
Нужно переименовать переменную

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/storage/UserDbStorageIT.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Test
+    void update_updates_fields() {
+        User u = userStorage.create(new User(0L, "a@a", "a", "A", LocalDate.of(2000, 1, 1)));
```
Нужно переименовать переменную

desmoo0 committed:
Delete InMemory{...} files

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,34 +1,56 @@
+    private Set<Long> likes = new HashSet<>();
+
+    @AssertTrue(message = "Дата выхода не может быть ранее 1895-12-28")
+    private boolean isReleaseDateValid() {
```
Этот метод сейчас нигде не используется. Нужно создать [касмотрую аннотацию валидации](https://reflectoring.io/bean-validation-with-spring-boot/#:~:text=to%20none.-,A%20Custom%20Validator%20with%20Spring%20Boot,-If%20the%20available) и повесить ее на  поле releaseDate

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,34 +1,56 @@
 import java.util.Set;
 
 @Data
+@Builder
```
Билдер в коде нигде не используется - его нужно удалить

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,34 +1,56 @@
+import java.util.LinkedHashSet;
 import java.util.Set;
 
 @Data
```
Data чаще всего избыточная аннотация, она может таить в себе скрытые проблемы. Важно не генерировать код, который не будет потом использоваться. Вот отличная статья [про применение lombok](https://medium.com/nuances-of-programming/lombock-%D1%85%D0%BE%D1%80%D0%BE%D1%88%D0%B5%D0%B5-%D0%B8-%D0%BF%D0%BB%D0%BE%D1%85%D0%BE%D0%B5-%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-4bd9cfdc5dd5)

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,34 +1,56 @@
 @Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
```
Из всех добавленных здесь аннотаций, должны остаться только `@Getter` и `@Setter`, которые действительно используются в коде

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,34 +1,56 @@
\ No newline at end of file
+    private Mpa mpa;
+
+    @Valid
```
Сейчас `@Valid` бессмысленен - у Genre нет ни одной валидационной аннотации, значит каскадная проверка ничего не проверит

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Genre.java:
```diff
@@ -0,0 +1,15 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode(of = "id")
```
`@Data` уже генерит `equals/hashCode`, а ты сверху ещё делаешь `@EqualsAndHashCode(of = "id")` - получается дублирование и путаница. Нужно оставить только аннотации, которые действительно используются;

Так как жанры мы храним в `Set<Genre>`, важно правильно настроить equals/hashCode, чтобы не словить дубли или пропадающие элементы. Для этого нужно днлать сравнение только по id и лучше использовать `Integer`, а не `int`. Лучше задать `@EqualsAndHashCode(onlyExplicitlyIncluded = true)` с `@EqualsAndHashCode.Include` на поле id.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Mpa.java:
```diff
@@ -0,0 +1,13 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class Mpa {
+    private int id;
```
Давай заменим на Integer

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Mpa.java:
```diff
@@ -0,0 +1,13 @@
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
```
Давай оставим аннотации, который действительно используются

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -1,29 +1,64 @@
 @Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
Нужно оставить только те аннотации, которые действительно используются

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -1,29 +1,64 @@
+        this.friends = new HashSet<>();
+    }
+
+    public Set<Long> getFriends() {
```
Этот метод нигде не используется

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -1,29 +1,64 @@
+    }
+
+    @Override
+    public int hashCode() {
```
Переопределять `equals` и `hashCode` не нужно, так как теперь User хранится в базе. Давай удалим лишний код

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,28 @@
+@RestController
+@RequiredArgsConstructor
+public class GenreController {
+    private final GenreStorage storage;
```
Должен использоваться сервис, а не репозиторий

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,70 +1,97 @@
-        return filmStorage.findById(id)
-                .orElseThrow(() -> new NoSuchElementException("Фильм с id=" + id + " не найден"));
+    private void enrichRefs(Film film) {
+        film.setMpa(mpas.findById(film.getMpa().getId())
```
Вместо того, чтобы обращаться к хранилищам и дублировать валидацию полученного значения, нужно использовать сервис и вызывать уже готовый метод getById, который выбросит исключение, если объект не найден

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,70 +1,97 @@
+
+        Set<Genre> resolvedGenres = film.getGenres().stream()
+                .map(g -> {
+                    Genre found = genres.findById(g.getId())
```
Здесь нужно обращаться к сервису жанров

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,70 +1,97 @@
-        }
-        return filmStorage.update(film);
+    public void removeLike(Long filmId, Long userId) {
+        users.findById(userId).orElseThrow(() -> new NotFoundException("User " + userId));
```
Здесь нужно обращаться к сервису пользователей

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,70 +1,97 @@
 
-        return filmStorage.create(film);
+    public void addLike(Long filmId, Long userId) {
+        users.findById(userId).orElseThrow(() -> new NotFoundException("User " + userId));
```
Здесь нужно обращаться к сервису пользователей

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,70 +1,97 @@
+    private final FilmStorage films;
+    private final GenreStorage genres;
+    private final MpaStorage mpas;
+    private final UserStorage users;
```
Из всех переменных класса должен остаться только одно хранилище - FilmsStorage, все остальные параметры долдны быть сервисами, чтобы избежать дублирований валидации

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,25 @@
+
+@Service
+@RequiredArgsConstructor
+public class GenreService {
```
Класс должен использоваться

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,25 @@
+@Service
+@RequiredArgsConstructor
+public class GenreService {
+    private final @Qualifier("genreDbStorage") GenreStorage storage;
```
Квалифаер нужно удалить

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,99 +1,79 @@
+    }
 
+    private void normalizeName(User user) {
         if (user.getName() == null || user.getName().isBlank()) {
```
Этот код нужно перенести в конструктор класса User, так как это часть создания валидного пользования. (Сейчас эта проверка может не работать, потому что у пользователя создается множество контрукторов с помощью аннотаций lombok,и jackson просто не понимает, какой констроктор ему использовать)

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -0,0 +1,213 @@
+    public boolean containsKey(Long id) {
+        final String sql = "SELECT COUNT(1) FROM FILM WHERE ID = ?";
+        Integer count = jdbc.queryForObject(sql, Integer.class, id);
+        return count != null && count > 0;
```
`COUNT(1)` всегда возвращает ровно одну строку: если записей нет - это будет 0, а не null, поэтому проверка `count != null` здесь лишняя.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+    @Override
+    public boolean containsKey(Long id) {
+        Integer cnt = jdbc.queryForObject("SELECT COUNT(1) FROM USERS WHERE ID=?", Integer.class, id);
+        return cnt != null && cnt > 0;
```
Проверку `cnt != null` нужно удалить (смотри FilmStorage)

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+            u.setEmail(rs.getString("EMAIL"));
+            u.setLogin(rs.getString("LOGIN"));
+            u.setName(rs.getString("NAME"));
+            Date d = rs.getDate("BIRTHDAY");
```
Нужно переименовать переменную

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+        u.setEmail(rs.getString("EMAIL"));
+        u.setLogin(rs.getString("LOGIN"));
+        u.setName(rs.getString("NAME"));
+        java.sql.Date birthdayDate = rs.getDate("BIRTHDAY");
```
` java.sql.` нужно перенести в импорты

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryGenreStorage.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.*;
+
+@Component
+public class InMemoryGenreStorage implements GenreStorage {
```
Нужно удалить inmemory реализацию

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -3,16 +3,14 @@
+import java.util.stream.Collectors;
 
 @Component
 public class InMemoryUserStorage implements UserStorage {
```
Нужно удалить inmemory реализацию

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/service/UserServiceTest.java:
```diff
@@ -1,67 +1,74 @@
-        userService = new UserService(new InMemoryUserStorage(), new InMemoryFilmStorage());
+    void setUp() {
+        // 1. Создаем наше "игрушечное" хранилище
+        UserStorage userStorage = new InMemoryUserStorage();
```
Так как inmemory хранилище мы удалим, можно [замокать репозиторий](https://habr.com/ru/companies/otus/articles/780090/), чтобы тесты продолжали работать

desmoo0 committed:
Refactoring & bugfixes

FilmStorage & GenreStorage & MpaStorage & UserStorage - были перемещены в /storage/function/ для большей удобности и читаемости.

desmoo0 committed:
Refactoring code

desmoo0 committed:
Fix unused imports

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java:
```diff
@@ -0,0 +1,127 @@
+    @Override
+    public Optional<User> findById(Long userId) {
+        final String sqlQuery = "SELECT ID, EMAIL, LOGIN, NAME, BIRTHDAY FROM USERS WHERE ID=?";
+        List<User> userList = jdbcTemplate.query(sqlQuery, (resultSet, rowNum) -> new User(
```
```suggestion
        List<User> userList = jdbcTemplate.query(sqlQuery, userRowMapper, userId);
```
